packages/shared/src/mergeChainedOptions.ts

```diff
@@ -7,43 +7,74 @@ import { isFunction, isPlainObject } from './utils';
 
 export type Falsy = false | null | undefined;
 
-export function mergeChainedOptions<T, U>(params: {
+export function mergeChainedOptions<T, U = unknown, F = Falsy>(params: {
   defaults: T;
-  options?: ChainedConfig<T> | Falsy;
+  options?: ChainedConfig<T> | F;
   mergeFn?: typeof Object.assign;
+  isFalsy?: (v: T | ChainedConfig<T> | F) => v is F;
 }): T;
-export function mergeChainedOptions<T, U>(params: {
+export function mergeChainedOptions<T, U, F = Falsy>(params: {
   defaults: T;
-  options: ChainedConfigWithUtils<T, U> | Falsy;
+  options: ChainedConfigWithUtils<T, U> | F;
   utils: U;
   mergeFn?: typeof Object.assign;
+  isFalsy?: (v: T | ChainedConfigWithUtils<T, U> | F) => v is F;
 }): T;
-export function mergeChainedOptions<T, U>(params: {
+export function mergeChainedOptions<T, U, F = Falsy>(params: {
   defaults: T;
-  options: ChainedConfigCombineUtils<T, U> | Falsy;
+  options: ChainedConfigCombineUtils<T, U> | F;
   utils: U;
   mergeFn?: typeof Object.assign;
   useObjectParam?: true;
+  isFalsy?: (v: T | ChainedConfigCombineUtils<T, U> | F) => v is F;
 }): T;
-export function mergeChainedOptions<T extends Record<string, unknown>>({
+export function mergeChainedOptions<T, U = unknown, F = Falsy>({
   defaults,
   options,
   utils,
   mergeFn = Object.assign,
   useObjectParam,
+  isFalsy = (
```

Comment:
Can we simplify this type declaration, complex types are difficult to maintain 😂

Reply:
ya. since parameters type of implementation signature does not that matter, i've roll the complex types back to `unknown` so that we can pay more focus on the overload signatures.

Reply:
👍